clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonInstanceTypeCachingAgent.java

```diff
@@ -0,0 +1,196 @@
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+    try {
+      Set<String> matchingAccounts = accountCredentialsRepository.getAll()
+        .stream()
+        .filter(AmazonCredentials.class::isInstance)
+        .map(AmazonCredentials.class::cast)
+        .filter(ac -> ac.getRegions().stream().anyMatch(r -> region.equals(r.getName())))
+        .map(AccountCredentials::getName)
+        .collect(Collectors.toSet());
+
+      if (matchingAccounts.isEmpty()) {
+        return new DefaultCacheResult(Collections.emptyMap());
+      }
+
+      CacheData metadata = providerCache.get(getAgentType(), "metadata", RelationshipCacheFilter.none());
+      MetadataAttributes metadataAttributes = null;
+      if (metadata != null) {
+        metadataAttributes = objectMapper.convertValue(metadata.getAttributes(), MetadataAttributes.class);
+      }
+
+      Set<String> instanceTypes = null;
+      if (metadataAttributes != null && metadataAttributes.etag != null && metadataAttributes.cachedInstanceTypes != null) {
+
+        //we have enough from a previous request to not re-request if the etag is unchanged..
+        HttpResponse headResponse = httpClient.execute(pricingHost, new HttpHead(pricingUri));
+        EntityUtils.consumeQuietly(headResponse.getEntity());
+        if (headResponse.getStatusLine().getStatusCode() != 200) {
+          throw new Exception("failed to read instance type metadata for " + region + ": " + headResponse.getStatusLine().toString());
+        }
+
+        Optional<String> etag = getEtagHeader(headResponse);
+
+        if (etag.filter(metadataAttributes.etag::equals).isPresent()) {
+          instanceTypes = metadataAttributes.cachedInstanceTypes;
+        }
+      }
+      if (instanceTypes == null) {
+        HttpResponse getResponse = httpClient.execute(pricingHost, new HttpGet(pricingUri));
+        if (getResponse.getStatusLine().getStatusCode() != 200) {
+          EntityUtils.consumeQuietly(getResponse.getEntity());
+          throw new Exception("failed to read instance type data for " + region + ": " + getResponse.getStatusLine().toString());
+        }
+        Optional<String> etag = getEtagHeader(getResponse);
+        HttpEntity entity = getResponse.getEntity();
+        instanceTypes = fromStream(entity.getContent());
+        EntityUtils.consumeQuietly(entity);
+        if (etag.isPresent()) {
+          metadataAttributes = new MetadataAttributes();
+          metadataAttributes.etag = etag.get();
+          metadataAttributes.cachedInstanceTypes = new HashSet<>(instanceTypes);
+          metadata = new DefaultCacheData("metadata", objectMapper.convertValue(metadataAttributes, ATTRIBUTES), Collections.emptyMap());
+        } else {
+          metadata = null;
+        }
+      }
+      Map<String, Collection<String>> evictions = new HashMap<>();
+      Map<String, Collection<CacheData>> cacheResults = new HashMap<>();
+      List<CacheData> instanceTypeData = new LinkedList<>();
```

Comment:
any particular reason to use a LinkedList over an ArrayList here? I could see an advantage if we were to remove random elements, but it seems we only ever `add` to the list

Reply:
honestly I figured that it is just an append-only list of somewhat unknown size that is then only iterated over afterwords not random-accessed so linked list would make sense